### PR TITLE
Konnect entities: `KongPlugin` reconciler

### LINKED ISSUE #530: Konnect entities: KongPlugin reconciler
## Problem Statement

https://github.com/Kong/gateway-operator/pull/513 introduced the reconciler for the new `KongPluginBinding` CRD. Since we want to allow the legacy approach based on annotations as well, we need to introduce a new `KongPlugin` reconciler that watches for plugin annotations on entities and creates the underlying corresponding `KongPluginBinding`.

## Acceptance Criteria

- [ ] The plugin annotations of an object ar reconciled, and a corresponding `KongPluginBinding` is created

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:

**Which issue this PR fixes**

Fixes #530 

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -57,6 +57,9 @@
 - The `DataPlaneKonnectExtension` CRD has been introduced. Such a CRD can be attached
   to a `DataPlane` via the extensions field to have a konnect-flavored `DataPlane`.
   [#453](https://github.com/Kong/gateway-operator/pull/453)
+- Labeling `KongService`s with the `konghq.com/plugins` label results in the creation
+  of a KongPluginBInding resources, that gets reconciled and created in Konnect.
```
Comment (czeslavo) [inline CHANGELOG.md]:
```suggestion
  of a KongPluginBinding resource, that gets reconciled and created in Konnect.
```

File: modules/manager/controller_setup.go [hunk 2]
```diff
@@ -382,6 +384,13 @@ func SetupControllers(mgr manager.Manager, c *Config) (map[string]ControllerDef,
 			}
 			controllers[name] = controller
 		}
+		controllers[KongPluginControllerName] = ControllerDef{
+			Enabled: c.KonnectControllersEnabled,
+			Controller: konnect.NewKongPluginReconciler(
+				c.DevelopmentMode,
+				mgr.GetClient(),
+			),
+		}
```
Comment (czeslavo) [inline modules/manager/controller_setup.go]: Please add the reconciler to the map defined above: `konnectControllers`. It was added to prevent runtime overrides in case of duplicate map keys.

File: test/integration/test_kongplugin.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/kong/gateway-operator/pkg/consts"
+	"github.com/kong/gateway-operator/test/helpers"
+
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+func TestKongPlugins(t *testing.T) {
```
Comment (pmalek) [inline test/integration/test_kongplugin.go:28]: WDYT about leaving a comment about the fact that this doesn't really sync anything to Konnect but is only testing the creation/deletion of KongPluginBindings?

File: modules/manager/controller_setup.go (hunk 2 above)
Comment (pmalek) [inline modules/manager/controller_setup.go]: Please update to the scaffolding used above to prevent duplication of controllers.

File: controller/konnect/reconciler_kongplugin.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+		).
+		Watches(
+			&configurationv1alpha1.KongService{},
+			handler.EnqueueRequestsFromMapFunc(r.mapKongServices),
+		).
+		Complete(r)
+}
+
+// Reconcile reconciles a KongPlugin object.
+func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:59]: Could we try splitting the `Reconcile` method into smaller functions? Now it's hard to quickly grasp what are the actions and their order from the reader perspective.

File: CHANGELOG.md (hunk 1 above)
Comment (pmalek) [reply CHANGELOG.md]: > ... creation of a KongPluginBInding resources, that gets reconciled and created in Konnect

I'd suggest rewording this as the binding object is not what's created in Konnect but rather a plugin instance is created, binding to objects as per the binding spec. WDYT?

File: test/integration/test_kongplugin.go [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+		return len(kongPluginBindingList.Items) == 0
+	}, 15*time.Second, time.Second)
+	require.Eventually(t, func() bool {
+		err := GetClients().MgrClient.Get(GetCtx(), client.ObjectKeyFromObject(proxyCachekongPlugin), proxyCachekongPlugin)
+		return k8serrors.IsNotFound(err)
+	}, 15*time.Second, time.Second)
+
+	t.Log("delete the kongservice and check it gets collected, as the kongPluginBinding finalizer should have been removed")
+	err = GetClients().MgrClient.Delete(GetCtx(), kongService)
+	require.NoError(t, err)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Delete(GetCtx(), kongService))
```

File: test/integration/test_kongplugin.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+			Name:      proxyCachekongPluginName,
+			Namespace: ns.Name,
+		},
+		PluginName: "proxy-cache",
+		Config: apiextensionsv1.JSON{
+			Raw: []byte(`{"response_code": [200], "request_method": ["GET", "HEAD"], "content_type": ["text/plain; charset=utf-8"], "cache_ttl": 300, "strategy": "memory"}`),
+		},
+	}
+	err = GetClients().MgrClient.Create(GetCtx(), proxyCachekongPlugin)
+	require.NoError(t, err)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Create(GetCtx(), proxyCachekongPlugin))
```

File: test/integration/test_kongplugin.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+		err := GetClients().MgrClient.List(GetCtx(), kongPluginBindingList, client.InNamespace(ns.Name))
+		if err != nil {
+			return false
+		}
+		if len(kongPluginBindingList.Items) != 1 {
+			return false
+		}
+		managedKPB = &kongPluginBindingList.Items[0]
+		return true
+	}, 30*time.Second, time.Second)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]: Can we extract those into `const`s and reuse them across this test?

I'd also suggest decreasing the tick time from 1s to something like 250ms or so. There's a bunch of them in this test and that (tick period) is time it takes for the first predicate execution, hence they add up to the minimal time execution ( glancing at this test, it seems there are already 8 invocations like this which makes this test take at the very least 8s).

WDYT?

File: test/integration/test_kongplugin.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+	t.Log("delete managed kongPluginBinding, then check it gets recreated")
+	err = GetClients().MgrClient.Delete(GetCtx(), managedKPB)
+	require.NoError(t, err)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Delete(GetCtx(), managedKPB))
```

File: test/integration/test_kongplugin.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+		},
+		Spec: configurationv1alpha1.KongServiceSpec{
+			KongServiceAPISpec: configurationv1alpha1.KongServiceAPISpec{
+				Name: lo.ToPtr(ksName),
+				URL:  lo.ToPtr("http://example.com"),
+			},
+		},
+	}
+	err = GetClients().MgrClient.Create(GetCtx(), kongService)
+	require.NoError(t, err)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Create(GetCtx(), kongService))
```

File: test/integration/test_kongplugin.go (hunk 6 above)
Comment (pmalek) [inline test/integration/test_kongplugin.go]:
```suggestion
	require.NoError(t, GetClients().MgrClient.Create(GetCtx(), proxyCachekongPlugin))
```

File: test/integration/test_kongplugin.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rateLimitingkongPluginName,
+			Namespace: ns.Name,
+		},
+		PluginName: "rate-limiting",
+		Config: apiextensionsv1.JSON{
+			Raw: []byte(`{"minute": 5, "policy": "local"}`),
+		},
+	}
+	err := GetClients().MgrClient.Create(GetCtx(), rateLimitingkongPlugin)
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]: Nit: You might consider putting `GetClients().MgrClient` into its own variable.

File: test/integration/test_kongplugin.go [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+			return false
+		}
+		if len(kongPluginBindingList.Items) != 1 {
+			return false
+		}
+		managedKPB = &kongPluginBindingList.Items[0]
+		return true
+	}, 30*time.Second, time.Second)
+
+	t.Log("delete managed kongPluginBinding, then check it gets recreated")
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]: Nit: add a name of PKB to the log entry.

File: test/integration/test_kongplugin.go [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+		if err != nil {
+			return false
+		}
+		if len(kongPluginBindingList.Items) > 0 {
+			return false
+		}
+		return true
+	}, 15*time.Second, time.Second)
+
+	t.Log("deploying an unmanaged KongPluginBinding resource")
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]: Log the KPB name.

File: test/integration/test_kongplugin.go [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+			return false
+		}
+		return len(kongPluginBindingList.Items) == 0
+	}, 15*time.Second, time.Second)
+	require.Eventually(t, func() bool {
+		err := GetClients().MgrClient.Get(GetCtx(), client.ObjectKeyFromObject(proxyCachekongPlugin), proxyCachekongPlugin)
+		return k8serrors.IsNotFound(err)
+	}, 15*time.Second, time.Second)
+
+	t.Log("delete the kongservice and check it gets collected, as the kongPluginBinding finalizer should have been removed")
```
Comment (pmalek) [inline test/integration/test_kongplugin.go]: Log the service name.

File: controller/konnect/reconciler_kongplugin.go [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+	for _, kongService := range kongServiceList.Items {
+		var pluginSlice []string
+
+		// in case no kongpluginbindings are referencing the kongservice, but it has the finalizer,
+		// we need to remove the finalizer.
+		if len(pluginBindingsByServiceName[kongService.Name]) == 0 {
+			if controllerutil.RemoveFinalizer(&kongService, consts.CleanupPluginBindingFinalizer) {
+				if err = r.Client.Update(ctx, &kongService); err != nil {
+					return ctrl.Result{}, err
+				}
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go]: You might consider using `.IsConflict(err)` to explicitly handle the conflict here. Users shouldn't not be concerned about these types of errors.

Example: https://github.com/Kong/gateway-operator/blob/dcb56adfdf81438a191b4a1533f565a08c6fb704/controller/konnect/reconciler_generic.go#L151-L156

File: controller/konnect/reconciler_kongplugin.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+		return ctrl.Result{}, err
+	}
+	for _, pluginBinding := range referencingBindingList.Items {
+		if pluginBinding.Spec.PluginReference.Name == kongPlugin.Name {
+			pluginBindings = append(pluginBindings, pluginBinding)
+		}
+	}
+
+	// TODO(mlavacca): So far we are supporting only KongService targets here. We need to implement
+	// the same logic for KongRoute, KongConsumer, and KongConsumerGroup as well.
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go:88]: Can we have an issue for this linked here?

File: controller/konnect/reconciler_kongplugin.go [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+// KongPluginReconciler reconciles a KongPlugin object.
+type KongPluginReconciler struct {
+	DevelopmentMode bool
+	Client          client.Client
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go]: These can be unexported given that we have a contructor that exposes those fields.

File: controller/konnect/watch_kongplugin.go [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+func (r *KongPluginReconciler) mapKongServices(ctx context.Context, obj client.Object) []reconcile.Request {
+	logger := log.GetLogger(ctx, "KongPlugin", r.DevelopmentMode)
+	kongService, ok := obj.(*configurationv1alpha1.KongService)
+	if !ok {
+		log.Error(logger, errors.New("cannot cast object to KongService"), "KongService mapping handler", obj)
+		return []ctrl.Request{}
+	}
+
+	requests := []ctrl.Request{}
+	if plugins, ok := kongService.Annotations["konghq.com/plugins"]; ok {
```
Comment (pmalek) [inline controller/konnect/watch_kongplugin.go]: I'd consider putting the `konghq.com/plugins` into a named const.

File: pkg/consts/konnect.go [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package consts
+
+const (
+	// CleanupPluginBindingFinalizer is the finalizer attached to entities that are properly
+	// referenced by KongPluginBindings, that should be cleaned up when the KongService
+	// gets deleted.
```
Comment (pmalek) [inline pkg/consts/konnect.go]: 1. I believe this comment should be a bit more explicit. What does "properly referenced" mean in here?
2. "that should be cleaned up when the KongService gets deleted":
  - I believe we can use a generic term like `resource` instead of KongService to prevent us forgetting to update this comment when other types of objects are supported.
  - What should clean this finalizer? I believe this is worth to be mentioned, otherwise users might think that they are somehow expected to clean this up.

**Review by pmalek — COMMENTED** [review]
Looks very good 👍 Left a couple of comments to make this a bit clearer to follow. Let me know what you think about those.

File: CHANGELOG.md (hunk 1 above)
Comment (mlavacca) [reply CHANGELOG.md]: Yep I agree, reworded.

File: modules/manager/controller_setup.go (hunk 2 above)
Comment (mlavacca) [reply modules/manager/controller_setup.go]: 👍 see 0efc11a3d37c9300f515aa0bc07f70e213b34de4

File: test/integration/test_kongplugin.go (hunk 3 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go:28]: sure, please see fa311fe77fd5ea8c5eebe6a836333f10a343f8db

File: modules/manager/controller_setup.go (hunk 2 above)
Comment (mlavacca) [reply modules/manager/controller_setup.go]: done in https://github.com/Kong/gateway-operator/commit/0efc11a3d37c9300f515aa0bc07f70e213b34de4

File: test/integration/test_kongplugin.go (hunk 10 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go]: yep, see db6de2b045694792c3f6f30193512c6fcb54765a

File: test/integration/test_kongplugin.go (hunk 7 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go]: yes, it makes sense. I've updated the tick time to 500 ms as 250 looked a bit too tight to me. See c5362d004ca9626c50b0b973215bf1ed7f0a4747

File: test/integration/test_kongplugin.go (hunk 11 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go]: 6ac46b88b163193a309eaaab096317cb6e743a28

File: test/integration/test_kongplugin.go (hunk 12 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go]: 6ac46b88b163193a309eaaab096317cb6e743a28

File: test/integration/test_kongplugin.go (hunk 13 above)
Comment (mlavacca) [reply test/integration/test_kongplugin.go]: 6ac46b88b163193a309eaaab096317cb6e743a28

File: controller/konnect/reconciler_kongplugin.go (hunk 14 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go]: it makes sense, here it is 375457f4017e5873c284f2652e217f0576535f2b

File: controller/konnect/reconciler_kongplugin.go (hunk 15 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go:88]: sure thing: 5692bc7ae19b5064b92c983ab21877e434a206a3

File: controller/konnect/watch_kongplugin.go (hunk 17 above)
Comment (mlavacca) [reply controller/konnect/watch_kongplugin.go]: this is a tricky one. I'd love to have all our annotations gathered in a common place that can be consumed by KIC, KGO, or even 3rd party tools (see https://github.com/kubernetes-sigs/ingress2gateway/blob/main/pkg/i2gw/providers/kong/consts.go#L25-L31). I think that kubernetes-configuration would be a great place to gather them, WDYT?

Comment (pmalek) [reply controller/konnect/watch_kongplugin.go]: That could be a long term goal but as an interim step I'd suggest to just use 1 const in this repo. We can then use an alias for easy migration and then migrate to a const from another repo (like kubernetes-configuration). WDYT?

File: pkg/consts/konnect.go (hunk 18 above)
Comment (mlavacca) [reply pkg/consts/konnect.go]: yep, I agree. here it is c4a737641026a56f76d2fc0d75a84d0c41187ce8

File: controller/konnect/reconciler_kongplugin.go (hunk 16 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go]: 2172e05b25fa599ca8d3cc034536b74e0248702b

File: controller/konnect/watch_kongplugin.go (hunk 17 above)
Comment (mlavacca) [reply controller/konnect/watch_kongplugin.go]: makes sense 033f86f350150c99a9439da6b31843223cf673fe

File: controller/konnect/reconciler_kongplugin.go [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+			handler.EnqueueRequestsFromMapFunc(r.mapKongPluginBindings),
+		).
+		Watches(
+			&configurationv1alpha1.KongService{},
+			handler.EnqueueRequestsFromMapFunc(r.mapKongServices),
+		).
+		Complete(r)
+}
+
+// Reconcile reconciles a KongPlugin object.
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:56]: I think it would be helpful for future readers to add a comment explaining briefly what is the responsibility of this reconciler - that might be not obvious it's meant to create `KongPluginBindings` rather than create plugins in Konnect.

File: controller/konnect/reconciler_kongplugin.go [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+							return true
+						}
+						return false
+					}) {
+					// then mark it for deletion, as the plugin is not referenced by the KongService anymore
+					pluginBindingsToDelete = append(pluginBindingsToDelete, pb)
+				}
+			}
+
+			// iterate over all the KondService annotations
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go]:
```suggestion
			// iterate over all the KongService annotations
```

File: test/integration/test_kongplugin.go [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+// TestKongPlugins is an integration that aims at testing the KongPluginBinding resources
+// are properly created and cleaned up when Konnect entities get annotated with KongPlugin names.
+// It also tests that unmanaged KongPluginBindings are properly handled when managed KongPluginBindings
+// are in play.
+// NOTE: this test does not test the Konnect integration. No resource created here will be pushed to Konnect.
+func TestKongPlugins(t *testing.T) {
```
Comment (czeslavo) [inline test/integration/test_kongplugin.go:28]: Loose suggestion: this could probably be an envtest to save some CI/local cycles.

File: controller/konnect/reconciler_kongplugin.go [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+		client.InNamespace(kongPlugin.Namespace),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, pluginBinding := range referencingBindingList.Items {
+		if pluginBinding.Spec.PluginReference.Name == kongPlugin.Name {
+			pluginBindings = append(pluginBindings, pluginBinding)
+		}
+	}
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go]: How about extracting this to a separate method `getPluginBindingsUsingKongPlugin`?

File: controller/konnect/reconciler_kongplugin.go [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+	// Group the PluginBindings by KongService name
+	pluginBindingsByServiceName := map[string][]configurationv1alpha1.KongPluginBinding{}
+	for _, pluginBinding := range pluginBindings {
+		if pluginBinding.Spec.Targets.ServiceReference == nil ||
+			pluginBinding.Spec.Targets.ServiceReference.Group != configurationv1alpha1.GroupVersion.Group ||
+			pluginBinding.Spec.Targets.ServiceReference.Kind != "KongService" {
+			continue
+		}
+		pluginBindingsByServiceName[pluginBinding.Spec.Targets.ServiceReference.Name] = append(pluginBindingsByServiceName[pluginBinding.Spec.Targets.ServiceReference.Name], pluginBinding)
+	}
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go]: How about extracting this to a separate `groupPluginBindingsByKongServiceName` function?

File: controller/konnect/reconciler_kongplugin.go [hunk 24]
```diff
@@ -0,0 +1,278 @@
...
+
+	pluginBindingsToDelete := []configurationv1alpha1.KongPluginBinding{}
+	// pluginReferenceFound represents whether the plugin is referenced by any KongService
+	var pluginReferenceFound bool
+	for _, kongService := range kongServiceList.Items {
+		var pluginSlice []string
+
+		// in case no kongpluginbindings are referencing the kongservice, but it has the finalizer,
+		// we need to remove the finalizer.
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:122]: How will this work if we only list indexed Services (the ones referenced by KongPluginBindings) according to TODO? Won't we need a separate reconciler for `KongService`s to handle their finalizers?

File: controller/konnect/reconciler_kongplugin.go [hunk 25]
```diff
@@ -0,0 +1,278 @@
...
+		if pluginBinding.Spec.Targets.ServiceReference == nil ||
+			pluginBinding.Spec.Targets.ServiceReference.Group != configurationv1alpha1.GroupVersion.Group ||
+			pluginBinding.Spec.Targets.ServiceReference.Kind != "KongService" {
+			continue
+		}
+		pluginBindingsByServiceName[pluginBinding.Spec.Targets.ServiceReference.Name] = append(pluginBindingsByServiceName[pluginBinding.Spec.Targets.ServiceReference.Name], pluginBinding)
+	}
+
+	// Get all the KongServices referenced by the KongPluginBindings
+	// TODO(mlavacca): use indexers instead of listing all KongServices
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:103]: Can we have an issue tracking this?

File: controller/konnect/reconciler_kongplugin.go [hunk 26]
```diff
@@ -0,0 +1,278 @@
...
+		client.InNamespace(kongPlugin.Namespace),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	pluginBindingsToDelete := []configurationv1alpha1.KongPluginBinding{}
+	// pluginReferenceFound represents whether the plugin is referenced by any KongService
+	var pluginReferenceFound bool
+	for _, kongService := range kongServiceList.Items {
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:118]: I'm afraid that if we leave this code as is and attempt to handle other entities (routes, consumers, groups) in this reconciler, it will get very hard to maintain considering the volume of code in a single function we have right now. Please try to extract the `Service`-specific code to separate method(s) to avoid that.

File: CHANGELOG.md [hunk 27]
```diff
@@ -64,6 +64,10 @@
 - Add `KongService`, `KongRoute`, `KongConsumer`, and `KongConsumerGroup` watchers
   in the `KongPluginBinding` reconciler.
   [#571](https://github.com/Kong/gateway-operator/pull/571)
+- Annotating `KongService`s with the `konghq.com/plugins` annotation results in
+  the creation of a managed `KongPluginBInding` resource, which is taken by the
```
Comment (pmalek) [inline CHANGELOG.md]:
```suggestion
  the creation of a managed `KongPluginBinding` resource, which is taken by the
```

File: controller/konnect/reconciler_kongplugin.go [hunk 28]
```diff
@@ -0,0 +1,278 @@
...
+								Name: kongPlugin.Name,
+							},
+						},
+					}
+					k8sutils.SetOwnerForObject(&kongPluginBinding, &kongPlugin)
+					if err = r.client.Create(ctx, &kongPluginBinding); err != nil {
+						return ctrl.Result{}, err
+					}
+					log.Debug(logger, "Managed KongPluginBinding created", kongPluginBinding)
+				}
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:224]: How about extracting this to a separate function `createKongPluginBindingForKongPlugin`? It could accept a struct with objects to be used as targets (in our case only a Service for now).

File: controller/konnect/reconciler_kongplugin.go [hunk 29]
```diff
@@ -0,0 +1,278 @@
...
+		}
+	} else {
+		if controllerutil.RemoveFinalizer(&kongPlugin, consts.PluginInUseFinalizer) {
+			if err = r.client.Update(ctx, &kongPlugin); err != nil {
+				return ctrl.Result{}, err
+			}
+			log.Debug(logger, "KongPlugin finalizer removed", kongPlugin)
+			return ctrl.Result{}, nil
+		}
+	}
```
Comment (czeslavo) [inline controller/konnect/reconciler_kongplugin.go:268]: How about extracting this to a separate method, e.g. `handleKongPluginFinalizer`?

File: controller/konnect/reconciler_kongplugin.go [hunk 30]
```diff
@@ -0,0 +1,278 @@
...
+		logger         = log.GetLogger(ctx, entityTypeName, r.developmentMode)
+	)
+
+	// Fetch the KongPlugin instance
+	var kongPlugin configurationv1.KongPlugin
+	if err := r.client.Get(ctx, req.NamespacedName, &kongPlugin); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go]: You can simplify this by using:

```suggestion
		return ctrl.Result{}, client.IgnoreNotFound(err)
```

File: controller/konnect/reconciler_kongplugin.go [hunk 31]
```diff
@@ -0,0 +1,278 @@
...
+	log.Debug(logger, "reconciling", kongPlugin)
+
+	// Get the pluginBindings that use this KongPlugin
+	pluginBindings := []configurationv1alpha1.KongPluginBinding{}
+	referencingBindingList := configurationv1alpha1.KongPluginBindingList{}
+	err := r.client.List(
+		ctx,
+		&referencingBindingList,
+		client.MatchingFields{
+			IndexFieldKongPluginBindingKongPluginReference: kongPlugin.Name,
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go]: According to the `ExtractValue` func in here: https://github.com/Kong/gateway-operator/blob/4731ea64685befe828bbdcdc648d6f5e11660271/controller/konnect/index_kongpluginbinding.go#L60-L70

it extract the value from the plugin binding objects as `<ns>/<name>`. Doesn't that mean that this won't match and we need to prepend the `kongPlugin.Name` with `kongPlugin.Namespace + "/"`?

File: controller/konnect/reconciler_kongplugin.go [hunk 32]
```diff
@@ -0,0 +1,278 @@
...
+		},
+		client.InNamespace(kongPlugin.Namespace),
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, pluginBinding := range referencingBindingList.Items {
+		if pluginBinding.Spec.PluginReference.Name == kongPlugin.Name {
```
Comment (pmalek) [inline controller/konnect/reconciler_kongplugin.go]: Do we actually need to do this? Doesn't using the index above guarantee us that we have only the requested plugin bindings (with matching ref name)?

File: test/integration/test_kongplugin.go (hunk 21 above)
Comment (pmalek) [reply test/integration/test_kongplugin.go:28]: Even though this is a good idea I believe we can continue with this integration test and add envtest based test in a separate PR (due to being pressed for time to release 1.4 on time).

We can leave a note here that this can be converted to envtest based test though. WDYT?

**Review by pmalek — COMMENTED** [review]
I believe this is a good starting point.

Ideally we'd split the `Reconcile()` func into smaller ones but that can also be done when support for other entities is added after this gets merged.

1 thing I'd like to see in here though is a sample manifest in `config/samples`.

File: controller/konnect/reconciler_kongplugin.go (hunk 31 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go]: yep, the integration test was broken indeed :). I added this indexing in a rush and I overlooked this aspect. Fixed in 285eb9c353c58691f0a4477a73ed27eefa694738

File: controller/konnect/reconciler_kongplugin.go (hunk 32 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go]: yep, removed in 285eb9c353c58691f0a4477a73ed27eefa694738

File: controller/konnect/reconciler_kongplugin.go (hunk 24 above)
Comment (mlavacca) [reply controller/konnect/reconciler_kongplugin.go:122]: Yes, In the tracking issue of KongService indexing I've added this detail: https://github.com/Kong/gateway-operator/issues/596

**Comment by mlavacca** [comment]
Regarding the comments related to splitting up the logic here into smaller pieces, my point of view is that this is the first PR that poses the basis to iterate on it. We already have multiple issues in place (https://github.com/Kong/gateway-operator/issues/583, https://github.com/Kong/gateway-operator/issues/596) that address the improvements on this side, and looking for a way now, in the very first iteration loop to optimize code readability and modularity as much as possible is out of scope IMO. Once we'll add the missing pieces (other Kong entities, entities indexing, etc.), the context will be broad enough to improve the code modularity. As of now, I think we should consider it as a starting point, as @pmalek already pointed out.

File: config/samples/konnect-kongservice-plugin-annotated.yaml
Comment (pmalek) [inline config/samples/konnect-kongservice-plugin-annotated.yaml:1]: 👍

File: controller/konnect/reconciler_kongplugin.go (hunk 22 above)
Comment (czeslavo) [reply controller/konnect/reconciler_kongplugin.go]: As we agreed, we're going to improve that in the upcoming PRs.

File: controller/konnect/reconciler_kongplugin.go (hunk 26 above)
Comment (czeslavo) [reply controller/konnect/reconciler_kongplugin.go:118]: As we agreed, we're going to improve that in the upcoming PRs.

File: controller/konnect/reconciler_kongplugin.go (hunk 29 above)
Comment (czeslavo) [reply controller/konnect/reconciler_kongplugin.go:268]: As we agreed, we're going to improve that in the upcoming PRs.

File: controller/konnect/reconciler_kongplugin.go (hunk 28 above)
Comment (czeslavo) [reply controller/konnect/reconciler_kongplugin.go:224]: As we agreed, we're going to improve that in the upcoming PRs.

File: controller/konnect/reconciler_kongplugin.go (hunk 23 above)
Comment (czeslavo) [reply controller/konnect/reconciler_kongplugin.go]: As we agreed, we're going to improve that in the upcoming PRs.

**Review by czeslavo — APPROVED** [review]

**Review by pmalek — APPROVED** [review]

**Comment by pmalek** [comment]
I've resolved conflicts after merging #593.

Should be able to be merged now.
